src/pip/_internal/operations/prepare.py

```diff
@@ -376,109 +376,105 @@ def _download_should_save(self):
-                    ". pip is being responsible and not assuming it "
-                    "can delete this. Please delete it and try again."
-                    .format(req, req.source_dir)
-                )
-
-            # Now that we have the real link, we can tell what kind of
-            # requirements we have and raise some more informative errors
-            # than otherwise. (For example, we can raise VcsHashUnsupported
-            # for a VCS URL rather than HashMissing.)
-            if self.require_hashes:
-                # We could check these first 2 conditions inside
-                # unpack_url and save repetition of conditions, but then
-                # we would report less-useful error messages for
-                # unhashable requirements, complaining that there's no
-                # hash provided.
-                if link.is_vcs:
-                    raise VcsHashUnsupported()
-                elif link.is_existing_dir():
-                    raise DirectoryUrlHashUnsupported()
-                if not req.original_link and not req.is_pinned:
-                    # Unpinned packages are asking for trouble when a new
-                    # version is uploaded. This isn't a security check, but
-                    # it saves users a surprising hash mismatch in the
-                    # future.
-                    #
-                    # file:/// URLs aren't pinnable, so don't complain
-                    # about them not being pinned.
-                    raise HashUnpinned()
-
-            hashes = req.hashes(trust_internet=not self.require_hashes)
-            if self.require_hashes and not hashes:
-                # Known-good hashes are missing for this requirement, so
-                # shim it with a facade object that will provoke hash
-                # computation and then raise a HashMissing exception
-                # showing the user what the hash should be.
-                hashes = MissingHashes()
+    def prepare_linked_requirement(
+        self,
+        req,  # type: InstallRequirement
+        parallel_builds=False,  # type: bool
+    ):
+        # type: (...) -> AbstractDistribution
+        """Prepare a requirement that would be obtained from req.link
+        """
+        assert req.link
+        link = req.link
+        self._log_preparing_link(req, link)
+        if link.is_wheel and self.wheel_download_dir:
+            # Download wheels to a dedicated dir when doing `pip wheel`.
+            download_dir = self.wheel_download_dir
+        else:
+            download_dir = self.download_dir
 
+        with indent_log():
+            self._ensure_link_req_src_dir(req, download_dir, parallel_builds)
             try:
                 local_file = unpack_url(
                     link, req.source_dir, self.downloader, download_dir,
-                    hashes=hashes,
-                )
+                    hashes=self._get_linked_req_hashes(req))
```

Comment:
suggestion: assign this outside the try block, and then keep the hashes=hashes here.

Reply:
I prefer it this way since there's no ambiguity what `_get_linked_req_hashes` returns.